# Description: don't use GtkHeaderBar decorations under Unity, they still
# create integration issues and are too differents
# Upstream: https://bugzilla.gnome.org/show_bug.cgi?id=762332
Index: gedit-3.18.3/gedit/gedit-window.c
===================================================================
--- gedit-3.18.3.orig/gedit/gedit-window.c
+++ gedit-3.18.3/gedit/gedit-window.c
@@ -409,6 +409,29 @@ gedit_window_tab_removed (GeditWindow *w
 	peas_engine_garbage_collect (PEAS_ENGINE (gedit_plugins_engine_get_default ()));
 }
 
+static gboolean
+in_desktop (const gchar *name)
+{
+	const gchar *desktop_name_list;
+	gchar **names;
+	gboolean in_list = FALSE;
+	gint i;
+
+	desktop_name_list = g_getenv ("XDG_CURRENT_DESKTOP");
+	if (!desktop_name_list)
+		return FALSE;
+
+	names = g_strsplit (desktop_name_list, ":", -1);
+	for (i = 0; names[i] && !in_list; i++)
+		if (strcmp (names[i], name) == 0) {
+			in_list = TRUE;
+			break;
+		}
+	g_strfreev (names);
+
+	return in_list;
+}
+
 static void
 gedit_window_class_init (GeditWindowClass *klass)
 {
@@ -480,6 +503,10 @@ gedit_window_class_init (GeditWindowClas
 			      0);
 
 	/* Bind class to template */
+	if (in_desktop ("Flashback"))
+		gtk_widget_class_set_template_from_resource (widget_class,
+		                                             "/org/gnome/gedit/ui/gedit-window-unity.ui");
+	else
 	gtk_widget_class_set_template_from_resource (widget_class,
 	                                             "/org/gnome/gedit/ui/gedit-window.ui");
 	gtk_widget_class_bind_template_child_private (widget_class, GeditWindow, titlebar_paned);
@@ -1099,10 +1126,12 @@ set_title (GeditWindow *window)
 		gedit_app_set_window_title (GEDIT_APP (g_application_get_default ()),
 		                            window,
 		                            "gedit");
+		if (!in_desktop ("Flashback")) {
 		gtk_header_bar_set_title (GTK_HEADER_BAR (window->priv->headerbar),
 		                          "gedit");
 		gtk_header_bar_set_subtitle (GTK_HEADER_BAR (window->priv->headerbar),
 		                             NULL);
+		}
 		gtk_header_bar_set_title (GTK_HEADER_BAR (window->priv->fullscreen_headerbar),
 		                          "gedit");
 		gtk_header_bar_set_subtitle (GTK_HEADER_BAR (window->priv->fullscreen_headerbar),
@@ -1203,10 +1232,12 @@ set_title (GeditWindow *window)
 				    window,
 				    main_title);
 
+	if (!in_desktop ("Flashback")) {
 	gtk_header_bar_set_title (GTK_HEADER_BAR (window->priv->headerbar),
 	                          title);
 	gtk_header_bar_set_subtitle (GTK_HEADER_BAR (window->priv->headerbar),
 	                             subtitle);
+	}
 	gtk_header_bar_set_title (GTK_HEADER_BAR (window->priv->fullscreen_headerbar),
 	                          title);
 	gtk_header_bar_set_subtitle (GTK_HEADER_BAR (window->priv->fullscreen_headerbar),
@@ -3560,6 +3591,13 @@ _gedit_window_fullscreen (GeditWindow *w
 		return;
 
 	sync_fullscreen_actions (window, TRUE);
+	/* Under unity since the headerbar is not a decoration it's not handled
+	   in an automatic way, we need to manual hide/show it so the behaviour
+	   is the correct one */
+	if (in_desktop ("Flashback")) {
+		gtk_widget_hide (window->priv->headerbar);
+		gtk_widget_hide (window->priv->side_headerbar);
+	}
 
 	/* Go to fullscreen mode and hide bars */
 	gtk_window_fullscreen (GTK_WINDOW (&window->window));
@@ -3574,6 +3612,11 @@ _gedit_window_unfullscreen (GeditWindow
 		return;
 
 	sync_fullscreen_actions (window, FALSE);
+	if (in_desktop ("Flashback")) {
+		gtk_widget_show (window->priv->headerbar);
+		if (gtk_widget_get_visible (window->priv->side_panel))
+			gtk_widget_show (window->priv->side_headerbar);
+	}
 
 	/* Unfullscreen and show bars */
 	gtk_window_unfullscreen (GTK_WINDOW (&window->window));
Index: gedit-3.18.3/gedit/resources/gedit.gresource.xml.in
===================================================================
--- gedit-3.18.3.orig/gedit/resources/gedit.gresource.xml.in
+++ gedit-3.18.3/gedit/resources/gedit.gresource.xml.in
@@ -15,6 +15,7 @@
     <file preprocess="xml-stripblanks">ui/gedit-highlight-mode-dialog.ui</file>
     <file preprocess="xml-stripblanks">ui/gedit-highlight-mode-selector.ui</file>
     <file preprocess="xml-stripblanks">ui/gedit-window.ui</file>
+    <file preprocess="xml-stripblanks">ui/gedit-window-unity.ui</file>
     <file preprocess="xml-stripblanks">ui/gedit-open-document-selector.ui</file>
     <file preprocess="xml-stripblanks">ui/gedit-statusbar.ui</file>
     <file>css/gedit-style.css</file>
Index: gedit-3.18.3/gedit/resources/ui/gedit-window-unity.ui
===================================================================
--- /dev/null
+++ gedit-3.18.3/gedit/resources/ui/gedit-window-unity.ui
@@ -0,0 +1,539 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<interface>
+  <!-- interface-requires gtk+ 3.10 -->
+  <template class="GeditWindow" parent="GtkApplicationWindow">
+    <property name="can_focus">False</property>
+    <property name="has_focus">False</property>
+    <property name="is_focus">False</property>
+    <child>
+      <object class="GtkBox">
+        <property name="visible">True</property>
+        <property name="orientation">vertical</property>
+    <child>
+      <object class="GtkPaned" id="titlebar_paned">
+        <property name="visible">True</property>
+        <property name="position" bind-source="hpaned" bind-property="position" bind-flags="bidirectional|sync-create"/>
+        <style>
+          <class name="gedit-headerbar-paned"/>
+        </style>
+        <child>
+          <object class="GtkHeaderBar" id="side_headerbar">
+            <property name="visible" bind-source="side_panel" bind-property="visible" bind-flags="sync-create"/>
+            <property name="show_close_button">False</property>
+            <property name="title" translatable="yes">Documents</property>
+            <style>
+              <class name="inline-toolbar"/>
+            </style>
+          </object>
+          <packing>
+            <property name="resize">False</property>
+            <property name="shrink">False</property>
+          </packing>
+        </child>
+        <child>
+          <object class="GtkHeaderBar" id="headerbar">
+            <property name="visible">True</property>
+            <property name="show_close_button">False</property>
+            <style>
+              <class name="inline-toolbar"/>
+            </style>
+            <child>
+              <object class="GtkMenuButton" id="open_button">
+                <property name="visible">True</property>
+                <property name="tooltip_text" translatable="yes">Open a file</property>
+                <property name="valign">center</property>
+                <property name="use_popover">True</property>
+                <style>
+                  <class name="text-button"/>
+                  <class name="image-button"/>
+                </style>
+                <child>
+                  <object class="GtkBox" id="open_button_box">
+                    <property name="visible">True</property>
+                    <property name="can_focus">False</property>
+                    <property name="has_focus">False</property>
+                    <property name="is_focus">False</property>
+                    <property name="spacing">6</property>
+                    <child>
+                      <object class="GtkLabel" id="open_button_name">
+                        <property name="label" translatable="yes">_Open</property>
+                        <property name="use-underline">True</property>
+                        <property name="visible">True</property>
+                        <property name="valign">baseline</property>
+                      </object>
+                    </child>
+                    <child>
+                      <object class="GtkImage" id="open_button_arrow">
+                        <property name="visible">True</property>
+                        <property name="valign">baseline</property>
+                        <property name="icon_name">pan-down-symbolic</property>
+                      </object>
+                    </child>
+                  </object>
+                </child>
+                <child internal-child="accessible">
+                  <object class="AtkObject" id="open_button_a11y">
+                    <property name="accessible-name" translatable="yes">Open</property>
+                    <property name="accessible-description" translatable="yes">Open a file</property>
+                  </object>
+                </child>
+              </object>
+              <packing>
+                <property name="pack_type">start</property>
+              </packing>
+            </child>
+            <child>
+              <object class="GtkButton" id="new_button">
+                <property name="visible">True</property>
+                <property name="can_focus">True</property>
+                <property name="can_default">True</property>
+                <property name="receives_default">False</property>
+                <property name="tooltip_text" translatable="yes">Create a new document</property>
+                <property name="valign">center</property>
+                <property name="action_name">win.new-tab</property>
+                <property name="image">new_tab_image</property>
+                <child internal-child="accessible">
+                  <object class="AtkObject" id="new_button_a11y">
+                    <property name="accessible-name" translatable="yes">New</property>
+                    <property name="accessible-description" translatable="yes">Create a new document</property>
+                  </object>
+                </child>
+              </object>
+            </child>
+            <child>
+              <object class="GtkMenuButton" id="gear_button">
+                <property name="visible">True</property>
+                <property name="valign">center</property>
+                <property name="use_popover">True</property>
+                <property name="image">menu_image</property>
+              </object>
+              <packing>
+                <property name="pack_type">end</property>
+              </packing>
+            </child>
+            <child>
+              <object class="GtkButton" id="save_button">
+                <property name="label" translatable="yes">_Save</property>
+                <property name="use-underline">True</property>
+                <property name="visible">True</property>
+                <property name="valign">center</property>
+                <property name="tooltip_text" translatable="yes">Save the current file</property>
+                <property name="action_name">win.save</property>
+                <child internal-child="accessible">
+                  <object class="AtkObject" id="save_button_a11y">
+                    <property name="accessible-name" translatable="yes">Save</property>
+                    <property name="accessible-description" translatable="yes">Save the current file</property>
+                  </object>
+                </child>
+              </object>
+              <packing>
+                <property name="pack_type">end</property>
+              </packing>
+            </child>
+          </object>
+          <packing>
+            <property name="shrink">False</property>
+          </packing>
+        </child>
+      </object>
+    </child>
+    <child>
+      <object class="GtkOverlay" id="fullscreen_overlay">
+        <property name="visible">True</property>
+        <property name="can_focus">False</property>
+        <property name="has_focus">False</property>
+        <property name="is_focus">False</property>
+        <child>
+          <object class="GtkBox" id="main_box">
+            <property name="visible">True</property>
+            <property name="orientation">vertical</property>
+            <child>
+              <object class="GtkPaned" id="hpaned">
+                <property name="visible">True</property>
+                <property name="can_focus">True</property>
+                <property name="has_focus">False</property>
+                <property name="is_focus">False</property>
+                <style>
+                  <class name="gedit-side-panel-paned"/>
+                </style>
+                <child>
+                  <object class="GtkBox" id="side_panel_box">
+                    <property name="visible" bind-source="side_panel" bind-property="visible" bind-flags="sync-create"/>
+                    <property name="orientation">vertical</property>
+                    <child>
+                      <object class="GeditMenuStackSwitcher" id="side_panel_inline_stack_switcher">
+                        <property name="visible">False</property>
+                        <property name="stack">side_panel</property>
+                        <style>
+                          <class name="gedit-side-panel-stack-switcher"/>
+                        </style>
+                      </object>
+                      <packing>
+                        <property name="expand">False</property>
+                        <property name="fill">False</property>
+                        <property name="position">0</property>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GtkStack" id="side_panel">
+                        <property name="visible">False</property>
+                      </object>
+                      <packing>
+                        <property name="expand">True</property>
+                        <property name="fill">True</property>
+                        <property name="position">1</property>
+                      </packing>
+                    </child>
+                  </object>
+                  <packing>
+                    <property name="resize">False</property>
+                    <property name="shrink">False</property>
+                  </packing>
+                </child>
+                <child>
+                  <object class="GtkBox" id="content_box">
+                    <property name="visible">True</property>
+                    <property name="orientation">vertical</property>
+                    <child>
+                      <object class="GtkPaned" id="vpaned">
+                        <property name="visible">True</property>
+                        <property name="can_focus">True</property>
+                        <property name="has_focus">False</property>
+                        <property name="is_focus">False</property>
+                        <property name="orientation">vertical</property>
+                        <style>
+                          <class name="gedit-bottom-panel-paned"/>
+                        </style>
+                        <child>
+                          <object class="GeditMultiNotebook" id="multi_notebook">
+                            <property name="visible">True</property>
+                          </object>
+                          <packing>
+                            <property name="resize">True</property>
+                            <property name="shrink">False</property>
+                          </packing>
+                        </child>
+                        <child>
+                          <object class="GtkBox" id="bottom_panel_box">
+                            <property name="visible" bind-source="bottom_panel" bind-property="visible" bind-flags="sync-create"/>
+                            <property name="orientation">horizontal</property>
+                            <style>
+                              <class name="gedit-bottom-panel"/>
+                            </style>
+                            <child>
+                              <object class="GtkBox" id="bottom_panel_stack_box">
+                                <property name="visible">True</property>
+                                <property name="orientation">vertical</property>
+                                <child>
+                                  <object class="GtkStack" id="bottom_panel">
+                                    <property name="visible">False</property>
+                                  </object>
+                                  <packing>
+                                    <property name="expand">True</property>
+                                    <property name="fill">True</property>
+                                    <property name="position">1</property>
+                                  </packing>
+                                </child>
+                                <child>
+                                  <object class="GeditNotebookStackSwitcher" id="bottom_panel_stack_switcher">
+                                    <property name="visible">True</property>
+                                    <property name="stack">bottom_panel</property>
+                                    <style>
+                                      <class name="gedit-bottom-panel-stack-switcher"/>
+                                    </style>
+                                  </object>
+                                  <packing>
+                                    <property name="expand">False</property>
+                                    <property name="fill">False</property>
+                                    <property name="position">2</property>
+                                  </packing>
+                                </child>
+                              </object>
+                              <packing>
+                                <property name="expand">True</property>
+                                <property name="fill">True</property>
+                                <property name="position">1</property>
+                              </packing>
+                            </child>
+                            <child>
+                              <object class="GtkBox" id="bottom_panel_sidebar">
+                                <property name="visible">True</property>
+                                <property name="orientation">vertical</property>
+                                <child>
+                                  <object class="GeditCloseButton" id="bottom_panel_close_button">
+                                    <property name="visible">True</property>
+                                    <property name="relief">none</property>
+                                    <property name="tooltip-text" translatable="yes">Hide panel</property>
+                                    <property name="action_name">win.bottom-panel</property>
+                                  </object>
+                                  <packing>
+                                    <property name="expand">False</property>
+                                    <property name="fill">False</property>
+                                    <property name="position">1</property>
+                                  </packing>
+                                </child>
+                              </object>
+                              <packing>
+                                <property name="expand">False</property>
+                                <property name="fill">False</property>
+                                <property name="position">2</property>
+                              </packing>
+                            </child>
+                          </object>
+                          <packing>
+                            <property name="resize">False</property>
+                            <property name="shrink">False</property>
+                          </packing>
+                        </child>
+                      </object>
+                      <packing>
+                        <property name="expand">True</property>
+                        <property name="fill">True</property>
+                        <property name="position">2</property>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GeditStatusbar" id="statusbar">
+                        <property name="visible">True</property>
+                        <child>
+                          <object class="GeditStatusMenuButton" id="line_col_button">
+                            <property name="visible">True</property>
+                            <property name="use_popover">True</property>
+                            <child internal-child="label">
+                              <object class="GtkLabel">
+                                <property name="width_chars">18</property>
+                                <property name="can_focus">False</property>
+                              </object>
+                            </child>
+                          </object>
+                          <packing>
+                            <property name="expand">False</property>
+                            <property name="fill">False</property>
+                            <property name="pack_type">end</property>
+                          </packing>
+                        </child>
+                        <child>
+                          <object class="GeditStatusMenuButton" id="tab_width_button">
+                            <property name="visible">True</property>
+                            <property name="use_popover">True</property>
+                          </object>
+                          <packing>
+                            <property name="expand">False</property>
+                            <property name="fill">False</property>
+                            <property name="pack_type">end</property>
+                          </packing>
+                        </child>
+                        <child>
+                          <object class="GeditStatusMenuButton" id="language_button">
+                            <property name="visible">True</property>
+                            <property name="use_popover">True</property>
+                          </object>
+                          <packing>
+                            <property name="expand">False</property>
+                            <property name="fill">False</property>
+                            <property name="pack_type">end</property>
+                          </packing>
+                        </child>
+                      </object>
+                      <packing>
+                        <property name="expand">False</property>
+                        <property name="fill">True</property>
+                        <property name="position">1</property>
+                      </packing>
+                    </child>
+                  </object>
+                  <packing>
+                    <property name="resize">True</property>
+                    <property name="shrink">True</property>
+                  </packing>
+                </child>
+              </object>
+              <packing>
+                <property name="expand">True</property>
+                <property name="fill">True</property>
+                <property name="position">1</property>
+              </packing>
+            </child>
+          </object>
+        </child>
+        <child type="overlay">
+          <object class="GtkEventBox" id="fullscreen_eventbox">
+            <property name="visible">True</property>
+            <property name="can_focus">False</property>
+            <property name="valign">start</property>
+            <child>
+              <object class="GtkRevealer" id="fullscreen_controls">
+                <property name="visible">True</property>
+                <property name="can_focus">False</property>
+                <property name="reveal_child">False</property>
+                <property name="valign">start</property>
+                <child>
+                  <object class="GtkHeaderBar" id="fullscreen_headerbar">
+                    <property name="visible">True</property>
+                    <child>
+                      <object class="GtkMenuButton" id="fullscreen_open_button">
+                        <property name="visible">True</property>
+                        <property name="tooltip_text" translatable="yes">Open a file dialog</property>
+                        <property name="valign">center</property>
+                        <property name="use_popover">True</property>
+                        <style>
+                          <class name="text-button"/>
+                          <class name="image-button"/>
+                        </style>
+                        <child>
+                          <object class="GtkBox" id="fullscreen_open_button_box">
+                            <property name="visible">True</property>
+                            <property name="can_focus">False</property>
+                            <property name="has_focus">False</property>
+                            <property name="is_focus">False</property>
+                            <property name="spacing">6</property>
+                            <child>
+                              <object class="GtkLabel" id="fullscreen_open_button_name">
+                                <property name="label" translatable="yes">Open</property>
+                                <property name="visible">True</property>
+                              </object>
+                            </child>
+                            <child>
+                              <object class="GtkImage" id="fullscreen_open_button_arrow">
+                                <property name="visible">True</property>
+                                <property name="icon_name">pan-down-symbolic</property>
+                              </object>
+                            </child>
+                          </object>
+                        </child>
+                        <child internal-child="accessible">
+                          <object class="AtkObject" id="fullscreen_open_button_a11y">
+                            <property name="accessible-name" translatable="yes">Open</property>
+                            <property name="accessible-description" translatable="yes">Open a file</property>
+                          </object>
+                        </child>
+                      </object>
+                      <packing>
+                        <property name="pack_type">start</property>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GtkButton" id="fullscreen_new_button">
+                        <property name="visible">True</property>
+                        <property name="can_focus">True</property>
+                        <property name="can_default">True</property>
+                        <property name="receives_default">False</property>
+                        <property name="tooltip_text" translatable="yes">Create a new document</property>
+                        <property name="valign">center</property>
+                        <property name="action_name">win.new-tab</property>
+                        <property name="image">fullscreen_new_tab_image</property>
+                        <child internal-child="accessible">
+                          <object class="AtkObject" id="fullscreen_new_button_a11y">
+                            <property name="accessible-name" translatable="yes">New</property>
+                            <property name="accessible-description" translatable="yes">Create a new document</property>
+                          </object>
+                        </child>
+                      </object>
+                    </child>
+                    <child>
+                      <object class="GtkButton" id="leave_fullscreen_button">
+                        <property name="visible">True</property>
+                        <property name="valign">center</property>
+                        <property name="tooltip_text" translatable="yes">Leave Fullscreen</property>
+                        <property name="action_name">win.leave-fullscreen</property>
+                        <property name="image">leave_fullscreen_image</property>
+                        <child internal-child="accessible">
+                          <object class="AtkObject" id="leave_fullscreen_button_a11y">
+                            <property name="accessible-name" translatable="yes">Leave Fullscreen</property>
+                            <property name="accessible-description" translatable="yes">Leave Fullscreen</property>
+                          </object>
+                        </child>
+                      </object>
+                      <packing>
+                        <property name="pack_type">end</property>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GtkMenuButton" id="fullscreen_gear_button">
+                        <property name="visible">True</property>
+                        <property name="valign">center</property>
+                        <property name="image">fullscreen_menu_image</property>
+                      </object>
+                      <packing>
+                        <property name="pack_type">end</property>
+                      </packing>
+                    </child>
+                    <child>
+                      <object class="GtkButton" id="fullscreen_save_button">
+                        <property name="label" translatable="yes">Save</property>
+                        <property name="visible">True</property>
+                        <property name="valign">center</property>
+                        <property name="tooltip_text" translatable="yes">Save the current file</property>
+                        <property name="action_name">win.save</property>
+                        <child internal-child="accessible">
+                          <object class="AtkObject" id="fullscreen_save_button_a11y">
+                            <property name="accessible-name" translatable="yes">Save</property>
+                            <property name="accessible-description" translatable="yes">Save the current file</property>
+                          </object>
+                        </child>
+                      </object>
+                      <packing>
+                        <property name="pack_type">end</property>
+                      </packing>
+                    </child>
+                  </object>
+                </child>
+              </object>
+            </child>
+          </object>
+        </child>
+      </object>
+    </child>
+      </object>
+    </child>
+  </template>
+  <object class="GtkImage" id="new_tab_image">
+    <property name="visible">True</property>
+    <property name="icon_name">tab-new-symbolic</property>
+  </object>
+  <object class="GtkImage" id="menu_image">
+    <property name="visible">True</property>
+    <property name="icon_name">open-menu-symbolic</property>
+  </object>
+  <object class="GtkImage" id="fullscreen_new_tab_image">
+    <property name="visible">True</property>
+    <property name="icon_name">tab-new-symbolic</property>
+  </object>
+  <object class="GtkImage" id="fullscreen_menu_image">
+    <property name="visible">True</property>
+    <property name="icon_name">open-menu-symbolic</property>
+  </object>
+  <object class="GtkImage" id="leave_fullscreen_image">
+    <property name="visible">True</property>
+    <property name="icon_name">view-restore-symbolic</property>
+  </object>
+  <object class="GtkSizeGroup">
+    <property name="mode">horizontal</property>
+    <widgets>
+      <widget name="side_headerbar"/>
+      <widget name="side_panel"/>
+    </widgets>
+  </object>
+  <object class="GtkSizeGroup">
+    <property name="mode">vertical</property>
+    <widgets>
+      <widget name="new_button"/>
+      <widget name="open_button"/>
+      <widget name="gear_button"/>
+      <widget name="save_button"/>
+      <widget name="open_button"/>
+    </widgets>
+  </object>
+  <object class="GtkSizeGroup">
+    <property name="mode">vertical</property>
+    <widgets>
+      <widget name="fullscreen_new_button"/>
+      <widget name="fullscreen_open_button"/>
+      <widget name="fullscreen_gear_button"/>
+      <widget name="fullscreen_save_button"/>
+      <widget name="fullscreen_open_button"/>
+      <widget name="leave_fullscreen_button"/>
+    </widgets>
+  </object>
+</interface>
+
